### DISCUSSION TITLE
MQTT Datensatz als JSON String

### DISCUSSION THREAD

**Discussion by wildewuzz** [tbnobody/OpenDTU#173] [Ideas]
Moin,
erstmal ganz dickes LOB ! Das funktioniert super. Nachdem meine DTU pro abgeraucht ist, habe ich geschaut was geht und dann dieses projekt entdeckt. Funktionierte auf Anhieb ! Danke !!!

Jetzt aber meine Frage. Habe das in mein Smarthome ( Symcon ) eingebunden. Das geht per MQTT auch easy. Ist es aber möglich die Daten als JSON String ( alles auf einmal, zumindest die einzelnen String ( 0-4) ) schicken ? Im moment kommen die ja einzeln

Gürße
Christian

**Comment by wildewuzz** [comment]
Hat sich erstmal erledigt. Nutze jetzt die API ! DANKE !

**Comment by tbnobody** [comment]
Hallo Christian,
die API ist für diesen Zweck auch besser geeignet. MQTT wurde nach dem Prinzip "Jeder Wert eine Topic" entwickelt. JSON Output geht eigentlich am Design vorbei.

**Reply by bwired-nl** [reply]
mhh to handle the data and store it in 1 write in a database its very easy to have 1 status topic with all topics and corresponding payload as a json. This is used very often.

**Reply by madmartin** [reply]
> mhh to handle the data and store it in 1 write in a database its very easy to have 1 status topic with all topics and corresponding payload as a json. This is used very often.

Yes, I confirm. for example the widespread "Tasmota" esp firmware sends a lot of (status-)information in json format. This gives several values (5-10) in one single MQTT message, and the whole status has 11 messages. If each of these values would have been send in a single message....

What about adding a config option "output status in json format" which can be enabled/disabled in addition to the existing status messages?
So everyone can get what he/she likes.

**Reply by bwired-nl** [reply]
great option, makes handling the mqtt output much easier

**Reply by madmartin** [reply]
see #225

**Comment by FrodoVDR** [comment]
Ich hätte mir auch eine json Ausgabe per MQTT gewünscht, dann hätte ich mir nicht erst eine json Ausgabe für influx basteln müssen.
Json lässt sich direkt in eine Influxdb schreiben und man hat dann eine Messreihe mit allen Werten bzw. 3 Messreihen bei meiner Anlage.
Wechselrichter HM-600 + 2x PV Modul.

Wie kommt man an die api denn heran, mir ist nirgendwo die passende URL aufgefallen.

**Reply by bwired-nl** [reply]
for example: if i have a json in a mqtt payload i can save this directly in a database. now i have write a lot of code to do it

**Reply by ce3rd** [reply]
API urls stehen unter anderem im Quelltext des javascripts http://IP/js/app.js

zb http://IP/api/livedata/status/

**Reply by madmartin** [reply]
> zb http://IP/api/livedata/status/

hab ich mal mit `curl` versucht abzurufen... da kommt zwar was, aber in einem mir unbekannten binären format.

**Reply by tbnobody** [reply]
remove the trailing slash....

**Reply by ce3rd** [reply]
hi,
spannend, bei mir klappts auch mim slash am end - im browser, oder curl zb: 

```
~# curl --silent http://192.168.141.137/api/livedata/status/

[{"serial":"1161XXXXXXX","name":"hm1500stadl1","data_age":5,"reachable":true,"producing":true,"0":{"Power":{"v":13.10000038,"u":"W"},"Voltage":{"v":232.3000031,"u":"V"},"Current":{"v":0.059999999,"u":"A"},"Power DC":{"v":13.70000076,"u":"W"},......
```
obiger output  ist gekürzt.

ich hol mir dann mit jq blöcke oder einzelne werte raus, zb block:

```
~# curl --silent http://192.168.141.137/api/livedata/status/ | jq '.[] | ."0"."Power"'

{
  "v": 12.19999981,
  "u": "W"
}

```

oder einzelne werte die ich oldschool im cacti vergraphe :-) (ich komm noch aus der snmp zeit):

```
:~# curl --silent http://192.168.141.137/api/livedata/status/ | jq '.[] | ."0"."Power".v'

12
```
(heute is es echt schiach hier ;-) )

**Comment by FrodoVDR** [comment]
Was ich aber bisher nicht gefunden habe wie man die Werte aus dem Phase 1 Fenster als json bekommt. Über MQTT werden sie geliefert, aber im /api/livedata/status/ sind sie nicht enthalten und auch alle anderen API Abfragen die ich in der /js/app.js gefunden habe liefern mir nicht die Werte des Wechselrichters.

**Reply by FrodoVDR** [reply]
http://ip/api/devinfo/status liefert zwar die Seriennummern und Firmware Version. Allerdings nicht die Netzspannung, Frequenz, Temperatur und soweiter.

**Reply by tbnobody** [reply]
> aber im /api/livedata/status/ sind sie nicht enthalten

Doch... Channel 0....

**Reply by FrodoVDR** [reply]
Vor lauter Werte und der doppelten Arrays hatte ich das übersehen. Jetzt komme ich an die Werte.
Danke
